Plugins/NativeCamera/Source/Android/NativeCameraImpl.cpp

```diff
@@ -51,35 +70,92 @@ namespace Babylon::Plugins
         return oesTexture;
     }
 
-    std::string Camera::Impl::GetCameraId(bool frontCamera)
+    Camera::Impl::CameraConfiguration Camera::Impl::GetCameraId(uint32_t maxWidth, uint32_t maxHeight, bool frontCamera)
     {
+        // Get the list of available cameras
         API24::ACameraIdList *cameraIds = nullptr;
         GET_CAMERA_FUNCTION(ACameraManager_getCameraIdList)(m_cameraManager, &cameraIds);
 
-        std::string cameraId{};
+        const char* bestCameraId{nullptr};
+        uint32_t bestPixelCount{0};
+        uint32_t bestDimDiff{0};
+        uint32_t bestWidth{0};
+        uint32_t bestHeight{0};
+        int32_t bestSensorOrientation{0};
+        bool foundExactMatch{false};
 
+        // Iterate over all of the cameras and find the one that has the best stream configuration
         for (int i = 0; i < cameraIds->numCameras; ++i)
         {
-            const char *id = cameraIds->cameraIds[i];
+            const char* id = cameraIds->cameraIds[i];
 
             API24::ACameraMetadata *metadataObj;
             GET_CAMERA_FUNCTION(ACameraManager_getCameraCharacteristics)(m_cameraManager, id, &metadataObj);
 
             API24::ACameraMetadata_const_entry lensInfo = {};
             GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_LENS_FACING, &lensInfo);
 
+            API24::ACameraMetadata_const_entry sensorOrientation = {};
+            GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_SENSOR_ORIENTATION, &sensorOrientation);
+
             auto facing = static_cast<API24::acamera_metadata_enum_android_lens_facing_t>(lensInfo.data.u8[0]);
+            if (facing != (frontCamera ? API24::ACAMERA_LENS_FACING_FRONT : API24::ACAMERA_LENS_FACING_BACK))
+            {
+                // Ignore cameras facing the wrong direction
+                continue;
+            }
+
+            // Get all available stream configurations supported by the camera
+            API24::ACameraMetadata_const_entry streamConfigurations;
+            GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_SCALER_AVAILABLE_STREAM_CONFIGURATIONS, &streamConfigurations);
+            // format of the data:
+            // 0: format
+            // 1: width
+            // 2: height
+            // 3: input?
+
+            for (uint32_t j = 0; j < streamConfigurations.count; j += 4) {
+                int32_t format = streamConfigurations.data.i32[j + 0];
+                int32_t width = streamConfigurations.data.i32[j + 1];
```

Comment:
On iOS this value always comes back in landscape orientation, so we had to swap width and height when the device was in portrait mode to properly apply the constraints.  Do you know if the same applies here?

Reply:
My understanding is that the width/height from the constraints should be interpreted as if they were given in the sensors natural orientation. The API on Android always gives you back a width that correlates with the longest side of the sensor (in its natural orientation in other words).

On the web if you do a max width of 1280 and a max height of 720 you may get back a stream that is 720 wide by 1280 height. Which at first glance seems like it's breaking the height constraints, but with my above assumption it makes sense. My logic here is doing the same thing as the web.

Reply:
Double checked and you're right, will update my PR to match this behavior.